documentation/Calls.md

```diff
@@ -0,0 +1,664 @@
+to F77ish procedures, so that
+a bare code address can serve to represent the effective argument.
+But targets that can only be called with an explicit interface
+have the option of using a "fat pointer" (or additional argument)
+to represent a dummy procedure closure so as
+to avoid the overhead of constructing and reclaiming a trampoline.
+Procedure descriptors can also support multiple code addresses.
+
+### Naming
+
+External subroutines and functions (R503) and `ENTRY` points (R1541)
+with `BIND(C)` (R808) have linker-visible names that are either explicitly
+specified in the program or determined by straightforward rules.
+The names of other F77ish external procedures should respect the conventions
+of the target architecture for legacy Fortran '77 programs; this is typically
+something like `foo_`.
+
+In other cases, however, we have fewer constraints on external naming,
+as well as some additional requirements and goals.
+
+Module procedures need to be distinguished by the name of their module
+and (when they have one) the submodule where their interface was
+defined.
+Note that submodule names are distinct in their modules, not hierarchical,
+so at most two levels of qualification are needed.
+
+Pure `ELEMENTAL` functions (15.8) must use distinct names for any alternate
+entry points used for packed SIMD arguments of various widths if we support
+calls to these functions in SIMD parallel contexts.
+There are already conventions for these names in `libpgmath`.
+
+The names of non-F77ish external procedures
+should be distinguished as such so that incorrect attempts to call or pass
+them with an implicit interface will fail to resolve at link time.
+Fortran 2018 explicitly enables us to do this with a correction to Fortran
+2003 in 4.3.4(5).
+
+Last, there must be reasonably permanent naming conventions used
+by the F18 runtime library for those unrestricted specific intrinsic
+functions (table 16.2 in 16.8) and extensions that can be passed as
+arguments.
+
+In these cases where external naming is at the discretion
+of the implementation, we should use names that are not in the C language
+user namespace, begin with something that identifies
+the current incompatible version of F18, the module, the submodule, and
+elemental SIMD width, and are followed by the external name.
+The parts of the external name can be separated by some character that
+is acceptable for use in LLVM IR and assembly language but not in user
+Fortran or C code, or by switching case
+(so long as there's a way to cope with extension names that don't begin
+with letters).
+
+In particular, the period (`.`) seems safe to use as a separator character,
+so a `Fa.` prefix can serve to isolate these discretionary names from
+other uses and to identify the earliest link-compatible version.
+For examples: `Fa.mod.foo`, `Fa.mod.submod.foo`, and (for an external
+subprogram that requires an explicit interface) `Fa.foo`.
```

Comment:
I would like to be clear on when the `Fa.` prefix would change. If we change f18 in a way that is link-incompatible, does that mean we would also change the prefix to `Fb.`? So that code compiled with a compiler from before that change won't link with code from the new version?

Reply:
Yes, that's right.  I'll add text.

Reply:
A minor consideration: numbers are easier to increment indefinitely than letters, and might make the purpose for their inclusion in the scheme slightly more obvious: F1 ... F10 ... vs. Fa ... Fz ...